xonsh/jobs.py

```diff
@@ -0,0 +1,199 @@
+"""
+Job control for the xonsh shell.
+"""
+import os
+import sys
+import time
+import signal
+import builtins
+from collections import namedtuple
+
+ProcProxy = namedtuple('ProcProxy', ['stdout', 'stderr'])
+"""
+A class representing a Python function to be run as a subprocess command.
+"""
+
+try:
+    _shell_tty = sys.stderr.fileno()
+except OSError:
+    _shell_tty = None
+
+_shell_pgrp = os.getpgrp()
+
+
+def _clear_dead_jobs():
+    to_remove = set()
+    for num, job in builtins.__xonsh_all_jobs__.items():
+        obj = job['obj']
+        if isinstance(obj, ProcProxy) or obj.poll() is not None:
+            to_remove.add(num)
+    for i in to_remove:
+        del builtins.__xonsh_all_jobs__[i]
+        if builtins.__xonsh_active_job__ == i:
+            builtins.__xonsh_active_job__ = None
+    if builtins.__xonsh_active_job__ is None:
+        _reactivate_job()
+
+
+def _reactivate_job():
+    if len(builtins.__xonsh_all_jobs__) == 0:
+        return
+    builtins.__xonsh_active_job__ = max(builtins.__xonsh_all_jobs__.items(),
+                                        key=lambda x: x[1]['started'])[0]
+
+
+_block_when_giving = [signal.SIGTTOU, signal.SIGTTIN,
+                      signal.SIGTSTP, signal.SIGCHLD]
+
+
+def _give_terminal_to(pgid):
+    # over-simplified version of:
+    #    give_terminal_to from bash 4.3 source, jobs.c, line 4030
+    # this will give the terminal to the process group pgid
+    if _shell_tty is not None:
+        oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
+        os.tcsetpgrp(_shell_tty, pgid)
+        signal.pthread_sigmask(signal.SIG_SETMASK, oldmask)
+
+
+def print_one_job(num):
+    """
+    Print a line describing job number ``num``.
+    """
+    job = builtins.__xonsh_all_jobs__[num]
+    act = '*' if num == builtins.__xonsh_active_job__ else ' '
+    status = job['status']
+    cmd = [' '.join(i) if isinstance(i, list) else i for i in job['cmds']]
```

Comment:
Is there any chance that `i` would ever be anything other than a list?  I can change it to be safe.